sw/snRuntime/CMakeLists.txt

```
@@ -6,20 +6,44 @@ cmake_minimum_required(VERSION 3.13)
 
 # Allow snRuntime to be built as a standalone library.
 if (CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
+    # Read SnitchUtilities
     list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/../cmake)
-    set(CMAKE_TOOLCHAIN_FILE toolchain-gcc CACHE STRING "Toolchain to use")
-
-    project(snRuntime LANGUAGES C ASM)
     include(SnitchUtilities)
-
+    # Create snRuntime project
+    project(snRuntime LANGUAGES C ASM)
     add_compile_options(-O3 -g -ffunction-sections)
 else()
     # Export package information to includer.
     set(SNRUNTIME_DIR ${CMAKE_CURRENT_BINARY_DIR} PARENT_SCOPE)
-    set(SNRUNTIME_INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/include PARENT_SCOPE)
+    set(SNRUNTIME_SRC_DIR ${CMAKE_CURRENT_SOURCE_DIR} PARENT_SCOPE)
+    set(SNRUNTIME_INCLUDE_DIRS 
+        ${CMAKE_CURRENT_SOURCE_DIR}/include 
+        ${CMAKE_CURRENT_SOURCE_DIR}/vendor 
+        PARENT_SCOPE)
+endif()
+
+# Default memory regions
+if(SNITCH_RUNTIME STREQUAL "snRuntime-cluster")
+    set(MEM_DRAM_ORIGIN "0x80000000" CACHE STRING "Base address of external memory")
+    set(MEM_DRAM_SIZE "0x80000000" CACHE STRING "Size of external memory")
+    set(MEM_L1_ORIGIN "0x100000" CACHE STRING "Base address of L1 memory")
+    set(MEM_L1_SIZE "128K" CACHE STRING "Size of L1 memory")
```

That is really the only thing that makes me a bit unhappy as this will automatically break if the TCDM is being relocated or changed in size. Do you see a way of getting around this? The information should be present from the previous boot stage and passed in `a1` (and if it isn't we should pass it). I am definitely okay with merging but this is a potential regression I am not super happy with. 
```suggestion
    set(MEM_L1_ORIGIN "0x100000" CACHE STRING "Base address of L1 memory")
    set(MEM_L1_SIZE "128K" CACHE STRING "Size of L1 memory")
```

---

Agree, removed all occurrences of `L1` in c255628 and added a simple memory allocator in ce65156. Global data now goes into DRAM and is shared amongst all cores. Private data can be used as either local (stack) or global with the `__thread` modifier. 

The `memory` example shows how the L1 allocation can be used. The not so beautiful thing is, that we need to store the pointer to the data globally and therefore this results in a write to DRAM and each time a core wants to access this data, it has to fetch the address stored in the pointer in DRAM again. There is currently no way to "exchange" small bytes between a cluster's cores without going off-chip.

https://github.com/pulp-platform/snitch/blob/1989408ef952d6fd617eaf5406e23f80439c2e6f/sw/examples/memory.c#L37-L45

One way for the user would be to allocate and then copy the pointer into its TCDM. `pp` can then be used to access the allocated chunk without fetching `p` from DRAM first.
```c
static void* p;
if(core_idx == 0) p = snrt_l1alloc(size);
snrt_barrier();
void* pp = p;
```

Is that feasible or do you see any way to have cluster-private data in TCDM wihout knowing its address at compile time?

---

I think your solution is the only work-around that came to my mind, apart from being maybe a bit hacky (although I don't think it's that bad) it should be fine.

What came to my mind is that the TCDM is pretty much the same as the shared memory in CUDA, so it could be interesting to see how `__shared__` is implemented there (definitely not saying we should go down that route). But I assume that comes with some significant effort in compiler and runtime/driver support :-).